skyrl-train/examples/mini_swe_agent/mini_swe_generator.py

```diff
@@ -0,0 +1,251 @@
+from skyrl_train.generators.utils import (
+    get_rollout_metrics,
+)
+
+
+class DefaultAgentWithReminder(DefaultAgent):
+    def get_observation(self, response: dict) -> dict:
+        """Execute the action and return the output."""
+        output = self.execute_action(self.parse_action(response))
+        observation = self.render_template(self.config.action_observation_template, output=output)
+        remaining = self.config.step_limit - self.model.n_calls
+
+        if remaining == 1:
+            observation = f"{observation}\nREMINDER: You only have 1 turn left. Please provide the final answer"
+        elif remaining > 1:
+            observation = f"{observation}\nREMINDER: You have {remaining} turns left to arrive at the solution."
+
+        self.add_message("user", observation)
+        return output
+
+
+@ray.remote(num_cpus=0.01)
+def init_and_run(instance, litellm_model_name, sweagent_config, generator_cfg, data_source, sampling_params):
+    from loguru import logger
+
+    model_config = sweagent_config.get("model", {})
+    # Use new sampling parameters
+    # Can also have custom sampling parameters per trajectory (ex: custom max tokens)
+    model_config.setdefault("model_kwargs", {}).update(sampling_params)
+    model = get_model(litellm_model_name, model_config)
+
+    agent = None
+    env = None
+    extra_info = None
+    result = None
+    reward = 0
+    error = None
+    try:
+        env = get_sb_environment(sweagent_config, instance, data_source)
+        agent = DefaultAgentWithReminder(model, env, **sweagent_config.get("agent", {}))
+        exit_status, result = agent.run(instance["problem_statement"])  # type: ignore[arg-type]
+    except Exception as e:
+        logger.error(f"Error processing instance {instance['instance_id']}: {e}", exc_info=True)
+        exit_status, result = type(e).__name__, str(e)
+        error = str(e)
+        extra_info = {"traceback": traceback.format_exc()}
+    finally:
+        path = Path(generator_cfg.miniswe_traj_dir)
+        path.mkdir(parents=True, exist_ok=True)
+        path = path / f"{instance['instance_id']}.json"
+        if agent is not None:
+            eval_error = None
+            try:
+                result = evaluate_trajectory(instance, result, sweagent_config, data_source)
+                reward = int(result["resolved"])
+                error = result["eval_error"]
+                if error:
+                    logger.debug(f"Error during evaluation {error}")
+            except Exception as e:
+                logger.debug(f"Error during evaluation {e}")
+                logger.debug(f"traceback: {traceback.format_exc()}")
+                eval_error = str(e)
+                error = str(e)
+
+            save_traj(agent, path, exit_status=exit_status, result=result, extra_info=extra_info, reward=reward, eval_error=eval_error)  # type: ignore[arg-type]
+
+        if env is not None:
+            # trigger cleanup explicitly
+            env.cleanup()
```

Comment:
We probably shouldn't call this explicitly, it will be called by the environment's `__del__` method, right? Also e.g. the LocalEnvironment doesn't have such a method, so the code also gets more generically applicable by omitting it.